dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -85,6 +88,31 @@ getAllDhallFiles baseDir = do
         Text.Megaparsec.errorBundlePretty err <>
         "... documentation won't be generated for this file"
 
+    extractTypeIfInSource :: Expr Void Import -> Maybe (Expr Void Import)
+    extractTypeIfInSource expr = do
+        name <- maybeNameInLet expr
+        typeOfName name
```

Comment:
One potential problem with this approach is that we might report the wrong type when we have variable shadowing. For example in

```
let x : T0 = y0
let x : T1 = y1
in x
```

we'd return `T0` when the actual type is `T1`.

I'm actually not sure how to solve this best. We probably have some code in `dhall-lsp-server` that already gets this right. I also think that this might be a good opportunity to try using the `subExpressions` traversal, since it's bottom-up.

Reply:
@sjakobi I haven't checkout `dhall-lsp-server` yet, but wouldn't work if we return the _last_ binding type that matches the name?

Reply:
> wouldn't work if we return the _last_ binding type that matches the name?

That works when the variable index is `0`. In Dhall, a variable like `x` is really a shorthand for `x@0`. In general, we could also encounter an expression like

```
let x : Bool = True
let x : Natural = 0
in x@1
```

`x@1` refers to he `Bool` `x` here.

I don't really expect that we'll encounter cases like this a lot, but if we do this right already, we might avoid some headaches later on.

Reply:
Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

Reply:
Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

Reply:
> Alright, and what about searching from the last binding name while the variable index != `0`? and keep decrementing?

Sounds good to me!

> I checked out `dhall-lsp-server` but it uses type-checking, and we are here just traversing the original AST from the source code.

Oh, I forgot. Sorry!

Reply:
Didn't noticed I sent the same message twice :laughing:

I pushed the change, please let me know what you think :)